pkg/collector/corechecks/system/gpu/nvidia/jetson.go

```diff
@@ -0,0 +1,490 @@
+func (c *JetsonCheck) sendTemperatureMetrics(sender aggregator.Sender, field string) error {
+	temperatureFields := c.regexes[regexTemperatureIdx].FindAllStringSubmatch(field, -1)
+	if len(temperatureFields) <= 0 {
+		return errors.New("could not parse temperature fields")
+	}
+
+	for i := 0; i < len(temperatureFields); i++ {
+		tempValue, err := strconv.ParseFloat(temperatureFields[i][tempValue], 64)
+		if err != nil {
+			return err
+		}
+		temperatureZoneTags := []string{fmt.Sprintf("zone:%s", temperatureFields[i][tempZone])}
+		sender.Gauge("nvidia.jetson.gpu.temp", tempValue, "", temperatureZoneTags)
+	}
+
+	return nil
+}
+
+func (c *JetsonCheck) sendVoltageMetrics(sender aggregator.Sender, field string) error {
+	voltageFields := c.regexes[regexVoltageIdx].FindAllStringSubmatch(field, -1)
+	if len(voltageFields) <= 0 {
+		return errors.New("could not parse voltage fields")
+	}
+
+	for i := 0; i < len(voltageFields); i++ {
+		voltageProbeTags := []string{fmt.Sprintf("probe:%s", voltageFields[i][voltageProbeName])}
+		instantVoltage, err := strconv.ParseFloat(voltageFields[i][currentVoltage], 64)
+		if err != nil {
+			return err
+		}
+		sender.Gauge("nvidia.jetson.gpu.vdd.instant", instantVoltage, "", voltageProbeTags)
+
+		averageVoltage, err := strconv.ParseFloat(voltageFields[i][averageVoltage], 64)
+		if err != nil {
+			return err
+		}
+		sender.Gauge("nvidia.jetson.gpu.vdd.average", averageVoltage, "", voltageProbeTags)
+	}
+
+	return nil
+}
+
+func (c *JetsonCheck) processTegraStatsOutput(tegraStatsOuptut string) error {
+	sender, err := aggregator.GetSender(c.ID())
+	if err != nil {
+		return err
+	}
+
+	var metricSenders = []sendMetric{
+		c.sendRAMMetrics,
+		c.sendIRAMMetrics,
+		c.sendSwapMetrics,
+		c.sendGpuUsageMetrics,
+		c.sendCPUUsageMetrics,
+		c.sendTemperatureMetrics,
+		c.sendVoltageMetrics,
+	}
+	for _, metricSender := range metricSenders {
+		err = metricSender(sender, tegraStatsOuptut)
+		if err != nil {
+			return err
+		}
+	}
+	sender.Commit()
+	return nil
+}
+
+// Run executes the check
+func (c *JetsonCheck) Run() error {
+	cmd := exec.Command(c.tegraStatsPath, c.commandOpts...)
+
+	// Parse the standard output for the stats
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return err
+	}
+	go func() {
+		in := bufio.NewScanner(stdout)
+		if in.Scan() {
+			// We only need to read one line
+			line := in.Text()
+			log.Debugf("tegrastats: %s", line)
+			if err = c.processTegraStatsOutput(line); err != nil {
+				_ = log.Error(err)
+			}
+		} else {
+			_ = log.Warnf("tegrastats did not produce any output")
+		}
+		err = cmd.Process.Signal(os.Kill)
+		if err != nil {
+			_ = log.Errorf("unable to stop %s check: %s", checkName, err)
+		}
+	}()
+
+	// forward the standard error to the Agent logger
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return err
+	}
+	go func() {
+		in := bufio.NewScanner(stderr)
+		for in.Scan() {
+			_ = log.Error(in.Text())
+		}
+	}()
+
+	if err := cmd.Start(); err != nil {
```

Comment:
We don't wait for the command to finish ? We should not send metrics after the check run method has ended.

Reply:
The agent is design to run checks per runner. We compute stats show in the status page related to running time for example. We also have a logic to avoid running a check that is already running (when run time it longer that the check interval). It will also mess with the scheduler where we spread check run to avoid CPU spikes.

The whole agent is design around the fact that check will block a runner. We also have long running checks (check that are run once and never exit). Those might be more suited to you use case.

Reply:
Having long running check is fine. We might want to document that using this check require increasing the number of runner but the agent is design to handle it. We could also increment the frequency (the check could run once every minute for example). The interval doesn't really matter as long as it's blocking and the documentation of the check is up to date.

For long running check you should check with the integration team for use case and why they went that way.